### DISCUSSION TITLE
Can't compile Neovim with custom g++

### DISCUSSION THREAD

**Discussion by Gullumluvl** [neovim/neovim#30281] [Q&A] [answered]
### Problem

Dear Neovim people,

I am on a CentOS 7.9 (end of life distribution) server with no admin rights, and the system GCC version (4.8) is incompatible with building Neovim.

So I installed GCC 14.2 and put it in my PATH as both `gcc` and `gcc-14.2`, as well as `g++` and `g++-14.2`.

Then in the Neovim source directory, I compile like this:

```bash
make CMAKE_INSTALL_PREFIX=$HOME/custom/gcc-14.2 \
    CMAKE_C_COMPILER=gcc-14.2 CMAKE_CXX_COMPILER=g++-14.2 \
    DEPS_CMAKE_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2 -DCMAKE_CXX_COMPILER=g++-14.2"
```

When compiling the "deps", the `DEPS_CMAKE_FLAGS` is taken into account, so I succeed in compiling the deps like that (otherwise it fails to build the deps).

However, I have the feeling that GCC 4.8.5 was used to compile `nvim` itself, as per the output of:

```bash
strings -a ./build/bin/nvim | grep 'GCC: ('
```

```
GCC: (GNU) 4.8.5 20150623 (Red Hat 4.8.5-44)
GCC: (GNU) 14.2.0
```

_How can I enforce the use of my local GCC ?_ (Do I have to edit the Makefile?)

To be clear, my `nvim` executable is functional, but I am not sure it is entirely safe to have it compiled in such an heterogeneous manner. Also, if in the future it ends up requiring a specific GCC, I'd like to know how to specify it.

Thanks in advance!

#### Alternatives I tried:

I tried to export `CC=gcc-14.2` and `CXX=g++-14.2` prior to running the make command (no change).

I tried to wrap with `CMAKE_EXTRA_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2 -DCMAKE_CXX_COMPILER=g++-14.2"` instead: this gives me the warning:

```
CMake Warning:
  Manually specified variables were not used by the project:
     CMAKE_CXX_COMPILER
``` 

Note that I have either run `make distclean` or entirely deleted the neovim folder before trying a new compilation.

### Steps to reproduce

```bash
make -j8 CMAKE_INSTALL_PREFIX=$HOME/custom/gcc-14.2 \
    CMAKE_C_COMPILER=gcc-14.2 CMAKE_CXX_COMPILER=g++-14.2 \
    DEPS_CMAKE_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2 -DCMAKE_CXX_COMPILER=g++-14.2"
```

### Expected behavior

```bash
strings -a ./build/bin/nvim | grep 'GCC: ('
```

should only show my custom GCC, that is:

```
GCC: (GNU) 14.2.0
```

-----

```bash
nvim -V1 -v
```

Should not indicate `Compilation: /usr/bin/cc ...` but instead `Compilation: ~/bin/gcc-14.2 ...`

### Neovim version (nvim -v)

0.10.1

### Vim (not Nvim) behaves the same?

not applicable

### Operating system/version

CentOS Linux release 7.9.2009 (Core)

### Terminal name/version

xterm/tmux/screen

### $TERM environment variable

screen-256color

### Installation

Build from git release

**Comment by zeertzjq** [comment]
You need to run `rm -rf build` first.

**Reply by Gullumluvl** [reply]
Yes, I did remove the entire neovim folder every time.

**Comment by dundargoc** [comment]
```
make distclean
make CMAKE_INSTALL_PREFIX=$HOME/custom/gcc-14.2 \
    CMAKE_EXTRA_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2"
    DEPS_CMAKE_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2"
```

**Comment by Gullumluvl** [comment, marked as answer]
I found out that the output of `strings -a build/bin/nvim` is not reflecting which compiler compiled the final executable.

When nvim is compiled with the flag `CMAKE_BUILD_TYPE=RelWithDebInfo`, then `nvim -V1 -v` shows a line for the compile command:

```
Compilation: /home/gullumluvl/bin/gcc-14.2 -O2 -g -Og -g -flto=auto [...]
```

And either one of the following calls successfully set the custom GCC version (using `CC` or using `CMAKE_C_COMPILER`):

```bash
make -j8 install \
    CMAKE_INSTALL_PREFIX=$HOME/custom/gcc-14.2 \
    CMAKE_BUILD_TYPE=RelWithDebInfo \
    CMAKE_EXTRA_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2" \
    DEPS_CMAKE_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2"
```

as suggested by @dundargoc , or simply:

```bash
make -j8 install \
    CMAKE_INSTALL_PREFIX=$HOME/custom/gcc-14.2 \
    CMAKE_BUILD_TYPE=RelWithDebInfo \
    CC=gcc-14.2 \
    DEPS_CMAKE_FLAGS="-DCMAKE_C_COMPILER=gcc-14.2"
```

Not sure why the output of `strings -a` also shows my system GCC, but anyway I'll be fine I guess.
